irods_testing_environment/test_runner.py

```diff
@@ -0,0 +1,130 @@
+# grown-up modules
+import logging
+
+# local modules
+from . import context
+from . import execute
+
+class test_runner:
+    """A class that manages a list of tests and can execute them on a managed container."""
+
+    def __init__(self, executing_container, tests=None):
+        """Constructor for `test_runner`.
+
+        The list of tests is not required to be initialized here. `add_test` will append a
+        test to the end of the test list.
+
+        Arguments:
+        executing_container -- the container on which tests will be executed
+        tests -- optional list of tests to use with this runner
+        """
+        # TODO: each test runner is tied to a single container... might need to abstract this out later
+        self.executor = executing_container
+        self.tests = tests or list()
+        #self.passed_tests = list()
```

Comment:
Presently, it looks as if everything in `tests` that does not appear in `failed_tests` passed when in fact some or all of these tests may not have executed at all. However, I feel like it's cleaner to only show the failed tests because it gives a stronger signal of something going wrong.

Any thoughts appreciated

Reply:
I agree.

If there are other cases (e.g. *tests not executed*, *tests skipped*, etc.) and you can detect them with confidence, then they should be placed in a different group.

Reply:
Consider this scenario...
 - single executor
 - list of 3 tests (`test_ihelp`, `blah` (should fail), `test_icd`)
 - 1 test passes, 1 test fails, the last test is skipped as a result

Here's what that might look like right now:
```
==== begin test run results ====
-----
results for [ubuntu-1804-postgres-1012_irods-catalog-provider_1]
        test list:
                [test_ihelp]
                [blah]
                [test_icd]
        failed tests:
                [blah]
        return code:[1]
-----
List of failed tests:
        blah
Return code:[1]
==== end of test run results ====
```

As you can see, it's very clear which test failed, but it's not clear that `test_icd` didn't run. So that's the discussion.

Here's what it would look like if we printed everything:
```
==== begin test run results ====
-----
results for [ubuntu-1804-postgres-1012_irods-catalog-provider_1]
        test list:
                [test_ihelp]
                [blah]
                [test_icd]
        passed tests:
                [test_ihelp]
        skipped tests:
                [test_icd]
        failed tests:
                [blah]
        return code:[1]
-----
List of failed tests:
        blah
Return code:[1]
==== end of test run results ====
```
This is too cluttered to me and I have to look a little closer to understand what happened.

What if we just add tests which are executed to one of two lists: passed or failed. At the end, we print 3 lists:

1. All tests that were supposed to be executed (total)
2. Tests that failed
3. Tests that do not appear in passed or failed (skipped/did not execute)

Here's what that might look like if a test fails and it skips something:
```
==== begin test run results ====
-----
results for [ubuntu-1804-postgres-1012_irods-catalog-provider_1]
        test list:
                [test_ihelp]
                [blah]
                [test_icd]
        skipped tests:
                [test_icd]
        failed tests:
                [blah]
        return code:[1]
-----
List of failed tests:
        blah
Return code:[1]
==== end of test run results ====
```

That's a little clearer to me, and the failed tests aren't drowned out too much. The tests not listed are implied as having run and passed. Under the covers, we are maintaining a list of passed and failed tests, but printing out failed tests and not-passed tests.

Reply:
I like this.

What does the output look like if the skip list and failed list are empty?

Reply:
I think what it does now is good (yes, including the smiley face) but also with the skipped tests item:
```
==== begin test run results ====
-----
results for [ubuntu-1804-postgres-1012_irods-catalog-provider_1]
        test list:
                [test_ihelp]
                [test_isysmeta]
                [test_icd]
        skipped tests:
        failed tests:
        return code:[0]
-----
All tests passed! :)
==== end of test run results ====
```
How's that look?

Also, as I was typing all of this, it occurred to me that the output should probably be provided in JSON format and placed in a separate file in addition to the log output. I'll make an issue for that.

Reply:
also a possibility - you don't have to print the entire list at the top... if you print only 'passed, skipped, failed'

oh, does this handle the 'fail fast' scenario?   so you'd have 'passed, skipped, failed, did_not_run'... or maybe 'did_not_run' is a subset of skipped?   hmm...

Reply:
Yes to all of that.

I think the current log output is pretty good. It's easy to read and doesn't have any of the additional noise that JSON brings in. I'm also curious to see how the JSON output will look. That will help us determine if the JSON version is the correct path forward instead of providing it as a knob.

Reply:
I actually used `--fail-fast` to generate the original output that I pasted above and demonstrate the issue, so it's being handled.

From where I'm standing, "skipped" is a detail specific to each `test_runner` implementation. For instance, the python test suite uses `unittest` which can indicate that it has skipped a test. That `test_runner` could parse its own output to determine whether something was skipped and add it to the list. I don't know what that would look like for Catch2 or whatever other kind of test comes along later.

Generically speaking, the base class *can* tell when something simply was not executed at all. So that's why I kind of grouped "skipped" and "did not run" together in my head. I think making the distinction might be interesting if we ever wanted a clean way to determine which tests are being skipped.

Reply:
Oooops, didn't paste the thing you suggested. Here's what that would look like:
```
==== begin test run results ====
-----
results for [ubuntu-1804-postgres-1012_irods-catalog-provider_1]
        passed tests:
                [test_ihelp]
        skipped tests:
                [test_icd]
        failed tests:
                [blah]
        return code:[1]
-----
List of failed tests:
        blah
Return code:[1]
==== end of test run results ====
```

Reply:
I think I agree. Not sure if skipped/did_not_run are worth splitting, but having the total list shown at the end in separate buckets seems best. Maybe we can add that distinction later?

Reply:
Latest commit makes this format true:
```
==== begin test run results ====
-----
results for [ubuntu-1804-postgres-1012_irods-catalog-provider_1]
        passed tests:
                [test_ihelp]
        skipped tests:
                [test_icd]
        failed tests:
                [blah]
        return code:[1]
-----
List of failed tests:
        blah
Return code:[1]
==== end of test run results ====
```

Does that seem okay for now?

Reply:
love it.

Reply:
Cooool. Let's continue this discussion in #47